server/src/main/java/io/crate/execution/dml/ObjectIndexer.java

```diff
@@ -0,0 +1,217 @@
+import org.apache.lucene.index.IndexableField;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import io.crate.execution.dml.Indexer.ColumnConstraint;
+import io.crate.execution.dml.Indexer.Synthetic;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.ColumnIdent;
+import io.crate.metadata.IndexType;
+import io.crate.metadata.Reference;
+import io.crate.metadata.ReferenceIdent;
+import io.crate.metadata.RelationName;
+import io.crate.metadata.RowGranularity;
+import io.crate.metadata.SimpleReference;
+import io.crate.sql.tree.ColumnPolicy;
+import io.crate.types.ArrayType;
+import io.crate.types.DataType;
+import io.crate.types.ObjectType;
+import io.crate.types.StorageSupport;
+
+public class ObjectIndexer implements ValueIndexer<Map<String, Object>> {
+
+    private final ObjectType objectType;
+    private final HashMap<String, ValueIndexer<Object>> innerIndexers;
+    private final ColumnIdent column;
+    private final Function<ColumnIdent, Reference> getRef;
+    private final RelationName table;
+    private final Reference ref;
+    private final Function<ColumnIdent, FieldType> getFieldType;
+    private final HashMap<String, DataType<?>> innerTypes;
+
+    @SuppressWarnings("unchecked")
+    public ObjectIndexer(RelationName table,
+                         Reference ref,
+                         Function<ColumnIdent, FieldType> getFieldType,
+                         Function<ColumnIdent, Reference> getRef) {
+        this.table = table;
+        this.ref = ref;
+        this.getFieldType = getFieldType;
+        this.getRef = getRef;
+        this.column = ref.column();
+        this.objectType = (ObjectType) ArrayType.unnest(ref.valueType());
+        this.innerIndexers = new HashMap<>();
+        this.innerTypes = new HashMap<>(objectType.innerTypes());
+        for (var entry : objectType.innerTypes().entrySet()) {
+            String innerName = entry.getKey();
+            DataType<?> value = entry.getValue();
+            ColumnIdent child = column.getChild(innerName);
+            Reference childRef = getRef.apply(child);
+            if (childRef.granularity() != RowGranularity.PARTITION) {
+                ValueIndexer<?> valueIndexer = value.valueIndexer(
+                    table,
+                    childRef,
+                    getFieldType,
+                    getRef
+                );
+                innerIndexers.put(entry.getKey(), (ValueIndexer<Object>) valueIndexer);
+            }
+        }
+    }
+
+    @Override
+    public void indexValue(@Nullable Map<String, Object> value,
+                           XContentBuilder xContentBuilder,
+                           Consumer<? super IndexableField> addField,
+                           Consumer<? super Reference> onDynamicColumn,
+                           Map<ColumnIdent, Indexer.Synthetic> synthetics,
+                           Map<ColumnIdent, Indexer.ColumnConstraint> checks) throws IOException {
```

Comment:
```suggestion
                           Map<ColumnIdent, Indexer.ColumnConstraint> toValidate) throws IOException {
```
Everywhere else is called `toValidate` and I think it's confusing to name it `checks` here.

Reply:
I think `checks` would actually communicate more of the semantics as it is more specific. I can address this in a follow up